src/lib/gadgets/bitwise.ts

```diff
@@ -37,18 +38,12 @@ function not(a: Field, length: number, checked: boolean = false) {
   }
 
   // create a bitmask with all ones
-  let allOnesF = new Field(2n ** BigInt(length) - 1n);
-
-  let allOnes = Provable.witness(Field, () => {
-    return allOnesF;
-  });
-
-  allOnesF.assertEquals(allOnes);
+  let allOnes = new Field(2n ** BigInt(length) - 1n);
 
   if (checked) {
     return xor(a, allOnes, length);
   } else {
-    return allOnes.sub(a);
+    return allOnes.sub(a).seal();
```

Comment:
It makes sure that the constraint created by the subtraction is added at this point. In my tests I found that the constraint system created by the unchecked version is empty.

That's because additions and subtractions are treated as a form of lazy constraint / AST, which are only turned into an actual constraint once the result is included in some multiplication or assertion or custom gate.

`seal()` creates a new variable that is asserted equal with the old variable, which causes snarky to concretize the AST into constraints. In the case of a gadget, I thought this was the better behavior, as it's normalized and not dependent on the remaining circuit where this subtraction will end up in the CS.

Reply:
The snarky version of `seal()` could be used for that! The o1js version not because it doesn't "seal" constants. The o1js version was intended more to prevent unnecessary large numbers of constraints when the same long AST is flushed to generic gates again and again. The function added in this PR that does what you descirbe @jspada is called `toVar()`